README.md
Neurogami::SwingSet
===================

by James Britt / Neurogami

http://www.neurogami.com


DESCRIPTION:
-------------------

Wraps some basic Java Swing components in a nicer Ruby API.


FEATURES/PROBLEMS:
-------------------

Very few.  There is a file that defines assorted basic Swing components (Frame, Panel, TextField)
using simple Ruby.  It makes it easier, and perhaps more 'Rubyish' to define Swing items.

Note that the code has references to some JRuby stuff you need to call rake using jruby:

    jruby -S rake gem



SYNOPSIS
------------

Install the gem. 

Neurogami gems are hosted on gems.neurogami.com

See http://www.neurogami.com/gems/

In the root of your project, run 

     swingset
  

This will copy over the swingset.rb and related files into the lib/ruby path of your project.

At the moment there are no options for other paths. 

You can then create a UI class like so:

      # foo_ui.rb
      require 'swingset'

      include  Neurogami::SwingSet::Core

      class FooFrame < Frame   

        FRAME_WIDTH = 500
        FRAME_HEIGHT =300 

        LABEL_HEIGHT = 20
        LABEL_WIDTH = 200

        attr_accessor :search_label, :search_value

        def initialize *args
          super
          self.minimum_width  = FRAME_WIDTH
          self.minimum_height = FRAME_HEIGHT
          base_font = Font.new("Lucida Grande", 0, 14)
          TextField.default_font = base_font 
          Label.default_font     = base_font 
          set_up_components
        end

        def set_up_components
          component_panel = Panel.new
          component_panel.background_color(255, 255, 255)
          component_panel.size(FRAME_WIDTH, FRAME_HEIGHT)
          
          # Assumes you have the MiG layout jar ...
          component_panel.layout = Java::net::miginfocom::swing::MigLayout.new("wrap 2")

          @search_label = Label.new do |l|
            l.minimum_dimensions(LABEL_WIDTH, LABEL_HEIGHT)
            l.text = "Search phrase 1"
          end

          @search_value = TextField.new do |t|
            t.minimum_dimensions(LABEL_WIDTH, LABEL_HEIGHT)
            t.text = "Neurogami"
          end

          component_panel.add @search_label, 'grow x'
          component_panel.add @search_value, "gap unrelated"


          add component_panel

        end

      end


REQUIREMENTS
-------------------

This code came out of changes to the Monkeybars library, such that one could define the View Swing
class using plain Ruby in place of a compiled Java class.

Writing raw Java/Swing code can get tedious, so these wrapper classes were created. Nobody likes tedium, except maybe Philip Glass.


INSTALL
-------------------

    sudo gem sources -a http://gems.neurogami.com
    sudo gem install swingset

or

    sudo gem i swingset ––source http://gems.neurogami.com


LICENSE
--------------

(The MIT License)

Copyright (c) 2010 James Britt

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


---

Feed yor head.


History.txt
== 0.2.2 / 2010-01-11

* Merged in changes from stepheneb to restore the use of Font inheritance from Swing
* Changed the Rakefile to use Mr. Bones 3.2.0, but also kept a copy of the previous Rakefile for people who do not want to update their bones install




spec/spec_helper.rb

require File.expand_path(
    File.join(File.dirname(__FILE__), %w[.. lib swingset]))

Spec::Runner.configure do |config|
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
end

# EOF


spec/swingset_spec.rb

require File.join(File.dirname(__FILE__), %w[spec_helper])

describe Swingset do
end

# EOF


lib/swingset_utils.rb
module Neurogami
  module SwingSet

    # :stopdoc:
    VERSION = '0.2.3'
    LIBPATH = ::File.expand_path(::File.dirname(__FILE__)) + ::File::SEPARATOR
    PATH = ::File.dirname(LIBPATH) + ::File::SEPARATOR
    # :startdoc:

    # Returns the version string for the library.
    #
    def self.version
      VERSION
    end

    # Returns the library path for the module. If any arguments are given,
    # they will be joined to the end of the libray path using
    # <tt>File.join</tt>.
    #
    def self.libpath( *args )
      args.empty? ? LIBPATH : ::File.join(LIBPATH, args.flatten)
    end

    # Returns the lpath for the module. If any arguments are given,
    # they will be joined to the end of the path using
    # <tt>File.join</tt>.
    #
    def self.path( *args )
      args.empty? ? PATH : ::File.join(PATH, args.flatten)
    end

    # Utility method used to rquire all files ending in .rb that lie in the
    # directory below this file that has the same name as the filename passed
    # in. Optionally, a specific _directory_ name can be passed in such that
    # the _filename_ does not have to be equivalent to the directory.
    #
    def self.require_all_libs_relative_to( fname, dir = nil )
      dir ||= ::File.basename(fname, '.*')
      search_me = ::File.expand_path(
                                     ::File.join(::File.dirname(fname), dir, '**', '*.rb'))

                                     Dir.glob(search_me).sort.each {|rb| require rb}
    end


    def self.copy_over path = 'lib/ruby'
      require 'fileutils'
      here = File.dirname(File.expand_path(__FILE__))
      warn "I am in #{here}"
      if File.exist?("#{path}/swingset.rb") || File.exist?("#{path}/swingset")
        warn "It seems that the swingset files already exist. Remove or rename them, and try again."
        exit
      end
    FileUtils.mkdir_p path unless File.exists? path
    FileUtils.cp_r "#{here}/swingset", path, :verbose =>  true
    FileUtils.cp_r "#{here}/swingset.rb", path, :verbose =>  true
    FileUtils.cp_r "#{here}/swingset_utils.rb", path, :verbose =>  true
    end

  end  # module Swingset
end


# EOF



lib/swingset/swingset.rb
module Neurogami
  module SwingSet
    module MiG
      module ClassMethods

        def mig_layout
          # Consider if this method should take an optional version string and 
          # do something (fail? fallback?) if the given version does not match the
          # version this methods wants to load.  At the least your code could then
          # fail early if SwingSet upgrades to an imcompatible MiG version.
          # Would need a real use case for this though.
          require 'miglayout-3.7.2-swing'
        end
      end

      def self.included(base)
        base.extend(MiG::ClassMethods)
      end

    end

    module Core
      module SwingConstants
       %w{
        BOTTOM
        CENTER
        EAST
        HORIZONTAL
        LEADING
        LEFT
        NEXT
        NORTH
        NORTH_EAST
        NORTH_WEST
        PREVIOUS
        RIGHT
        SOUTH
        SOUTH_EAST
        SOUTH_WEST
        TOP
        TRAILING
        VERTICAL
        WEST}.each do  |konst|
         class_eval "#{konst} = Java::javax::swing::SwingConstants::#{konst}"
        end
      end

      class Dimension
        def self.[](width, height)
          java::awt::Dimension.new width, height
        end
      end

      class ImageIcon
        def self.load  image_path 
          javax.swing.ImageIcon.new load_resource image_path
        end
      end


      class GroupLayout
        def self.get content_pane
          org.jdesktop.layout.GroupLayout.new content_pane 
        end
      end

      # A button wrapper
      # See http://xxxxxxxx to understand Swing buttons
      class Button < Java::javax::swing::JButton
        def initialize
          super
          yield self if block_given?
        end
      end


      class MenuBar < Java::javax.swing.JMenuBar
        def initialize
          super
          yield self if block_given?
        end

      end

      class MenuItem  < Java::javax.swing.JMenuItem
        def initialize
          super
          yield self if block_given?
        end
      end

      class Menu  < Java::javax.swing.JMenu
        def initialize
          super
          yield self if block_given?
        end
      end

      class Font < Java::java.awt.Font
      end

      # A label  wrapper
      # See http://xxxxxxxx to understand Swing labels
      class Label < Java::javax::swing::JLabel

        @@default_font = java::awt.Font.new "Lucida Grande", 0, 12

        def self.default_font= default_font
          @@default_font = default_font
        end

        def self.default_font
          @@default_font 
        end

        def initialize text=nil
          super
          self.text = text.to_s
          self.font = Label.default_font
          yield self if block_given?
        end

        def minimum_dimensions width, height
          self.minimum_size = java::awt::Dimension.new width, height
        end

        def prefered_dimensions width, height
          self.prefered_size = java::awt::Dimension.new width, height
        end
      end


      class TextField < Java::javax.swing.JTextField

        @@default_font = java::awt.Font.new "Lucida Grande", 0, 12

        def self.default_font= default_font
          @@default_font = default_font
        end

        def self.default_font
          @@default_font 
        end

        def initialize text = nil
          super
          self.text = text.to_s
          self.font = Label.default_font

          yield self if block_given?
        end

        def minimum_dimensions width, height
          self.minimum_size = java::awt::Dimension.new width, height
        end

        def prefered_dimensions width, height
          self.prefered_size = java::awt::Dimension.new width, height
        end

      end

      # A LayeredPane wrapper
      # See http://xxxx xxxx to understand Swing LayeredPanes 
      class LayeredPane < javax::swing.JLayeredPane

        def background_color red, blue, green
          self.background = java::awt::Color.new red.to_i, blue.to_i, green.to_i
        end

        def size width, height
          self.preferred_size =  java::awt::Dimension.new width, height
        end


        def add_ordered_components *components
          components.each do |c|
            self.add c
          end

          components.each do |c|
            self.moveToFront c
          end


        end
      end

      # A panel  wrapper
      # See http://xxxxxxxx to understand Swing panels
      class Panel < javax::swing.JPanel

        def background_color red, blue, green
          self.background = java::awt::Color.new red.to_i, blue.to_i, green.to_i
        end

        def size width, height
          self.preferred_size =  java::awt::Dimension.new  width, height
        end
      end


      # A frame  wrapper
      # See http://xxxxxxxx to understand Swing frames
      class Frame  < Java::javax::swing::JFrame
        attr_accessor :minimum_height, :minimum_width

        def initialize *args
          super
        end

        def define_minimum_size width, height
          self.minimum_size = java::awt::Dimension.new width, height 
        end

        def minimum_height= height
          define_minimum_size  @minimum_width.to_i, @minimum_height = height.to_i
        end

        def minimum_width= width
          define_minimum_size  @minimum_width = width.to_i, @minimum_height.to_i
        end
      end
    end
  end
end


examples/basic.rb
# Note: The goal is include some examples if using Neurogami::SwingSet.
# The orginal goal of SwingSet was for use with Monkeybars and Jimpanzee.
# Both of those look after thngs such as event handlers and the UI lifecycle.
# Standalone examples of SwingSet need to take care of that, though it does not
# offer any Rubyesque API for them.
#
# Still, the mechanics are not that hard for general examples.
#

# This is nasty.  The loading of the MiG layout jar should be wrapped somehow
# so that you can load it without having to know where it lives or the version number
# on the jar.  This way, with updates to SwingSet, code that calls a local version
# of MiG will simply get whatever version was shipped for that version of SwingSet and you
# don't have to go fix your code to update the MiG version number.


$:.unshift File.expand_path(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'swingset'

include  Neurogami::SwingSet::Core

class BasicFrame < Frame   
  include  Neurogami::SwingSet::MiG

  mig_layout

  FRAME_WIDTH  = 600
  FRAME_HEIGHT = 130

  LABEL_WIDTH  = 400
  LABEL_HEIGHT = 60

  # Make sure our components are available! 
  attr_accessor :default_button, :default_label

  def initialize *args
    super
    self.minimum_width  = FRAME_WIDTH
    self.minimum_height = FRAME_HEIGHT
    set_up_components
    default_close_operation = EXIT_ON_CLOSE 
  end

  def set_up_components
    component_panel = Panel.new

    # If we were clever we would define a method that took a  single hex value, like CSS.
    component_panel.background_color 255, 255, 255
    component_panel.size FRAME_WIDTH, FRAME_HEIGHT

    # This code uses the MiG layout manager.
    # To learn more about MiGLayout, see:
    #     http://www.miglayout.com/
    component_panel.layout = Java::net::miginfocom::swing::MigLayout.new "wrap 2"

    @default_label = Label.new do |l|
      l.font = Font.new "Lucida Grande", 0, 18
      l.minimum_dimensions LABEL_WIDTH, LABEL_HEIGHT
      l.text = "Neurogami::SwingSet rulez!"
    end

    @default_button = Button.new do |b|
      b.text = "Close"
    end

    # Add components to panel
    component_panel.add @default_label, "gap unrelated"
    component_panel.add @default_button, "gap unrelated"
    add component_panel

    @default_button.addActionListener lambda{ |e| default_button_clicked e}
  end

  def default_button_clicked event
    puts "Our button was clicked"
    java.lang.System.exit 0
  end

end


af = BasicFrame.new
af.set_location  500, 300
af.show

